### ISSUE TITLE
How should URL table relate to 3 data sources?

### ISSUE THREAD

**Issue by kat-kel** [medialab/spsm-database#4]
I'd like some help thinking about how to relate a table of URLs to the 3 data sources we used to alimenter la base. I've stored as much data as possible from the 3 original data sources in 3 tables. I've even re-requested the databases De Facto and Science Feedback to maximize the amount of original data we have, just in case some day that's useful for a new research question.

1. Condor
    - primary key (`condor.id`): incremental integer because Condor's dataset has duplicates of its own internal ID (`condor_url_rid`)
2. Science Feedback
    - primary key (`science_feedback.id`): ID of the "appearance" of a fact-checked URL in Science Feedback's database
3. De Facto
    - primary key (`de_facto.id`): ID of the article that fact-checks the URL in De Facto's database

None of the 3 datasets aggregates by the URL, so they risk having duplicated URLs. What we want to do is relate each unique primary key / entity in the datasets to one entity in the URL table. This would probably not be complicated if the URL table's primary key was the hash of the normalized URL (`normalized_url.normalized_url_id`) The reason this would not be complicated is because every entity in `condor`, `de_facto`, and `science_feedback` has 1 URL and 1 hash of that URL's normalized version.

However, we have been discussing augmenting the URL table so that the entity = a completed URL, which may or may not exist in the original 3 datasets. For example, if an entity in `condor` had the normalized URL `"zooplus.de"` and during the Twitter collection we used the entity's `condor.share_title` to develop the URL `"https://zooplus.de/title"`, (1) this new URL does not exist in table `condor`, and (2) we want a hash of this new URL to be the primary key of the URL table.

I think one way to fix this is to add rows to the `condor` table, rather than the URL table. This would mean that we would also use the new URL to derive values for the columns `condor.url_id` and `condor.normalized_url`. (Remember, the Condor dataset already has duplicated internal IDs, `condor.url_rid`, which is why the `condor` table primary key is a serial integer.) Then, we would build the URL table, as I've done below, from the `condor`, `de_facto`, and `science_feedback` tables' normalized URL hashes.

Even with that idea, do we think the relations I've outlined below would be useful for the type of requests we're anticipating? Do we think it'd be easy to filter the URL table based on parameters searched in one or multiple of the related 3 dataset tables? For example, in this relational framework, would it be easy for someone to get a list of all the URLs for which the Science Feedback rating is equal to 1.0 and/or the De Facto rating is equal to 1? If not, what relational framework would you propose?

![ER-Diagram_urltable](https://github.com/medialab/spsm-database/assets/63346582/0d9a56b0-cd49-4844-8bd5-ec008601cf8e)

**Comment by kat-kel** [comment]
I've a new idea to propose. What if we created a table that represents the URLs that we derived from the URLs in our 3 data sources? Below, you'll see a new framework that has a new table `completed_urls`, which relates to the table `condor` where `completed_urls.condor_table_id` = `condor.id`.* One entity in the table `condor` will relate to multiple entities in the table `completed_urls`. Because the entity of the `completed_url` table is a URL, it has a one-to-one relationship with the table `url`.

A question I have about this proposed solution is this: Should the `completed_urls` table redundantly copy information from the related `condor` table, for ease of querying? Or should we leave it bare, in case we some day want to use the table to store more URLs that we "complete" / modify from other data sources' URLs in the database?

![image](https://github.com/medialab/spsm-database/assets/63346582/fb3cecd8-0ef7-4486-865d-3ee480b86cbf)

*It's necessary to relate each completed URL to the ID that I've attributed to entities in the `condor` table because some entities in our dataset of completed URLs relate to a Condor entity that Facebook did not give an ID. 12 entities in the completed URLs dataset didn't have a corresponding Condor ID. To attribute each completed URL to an entity in the `condor` table, I used 2 SQL queries.

https://github.com/medialab/spsm-database/blob/076dbe4e65fb204d738549d0ccdbbfffce6957b1/src/build-database/build_tables/data_sources.py#L48-L83
